zipkin-local.js

```diff
@@ -0,0 +1,129 @@
+"use strict";
+const _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
+const {
+  THEME_OPTIONS,
+} = require("gatsby-plugin-newrelic-test/utils/constants");
+const { getCiData } = require("./utils/getCiData");
+const ciAttributes = getCiData();
+exports.__esModule = true;
+exports.stop = exports.create = void 0;
+
+const _zipkin = _interopRequireDefault(require("zipkin"));
+
+const _zipkinTransportHttp = require("zipkin-transport-newrelic");
+
+const _zipkinJavascriptOpentracing = _interopRequireDefault(
+  require("zipkin-javascript-opentracing")
+);
+const _nodeFetch = _interopRequireDefault(require("node-fetch"));
+let logger;
+let recorder;
+/**
+ * Create and return an open-tracing compatible tracer. See
+ * https://github.com/opentracing/opentracing-javascript/blob/master/src/tracer.ts
+ */
+
+const create = () => {
+  logger = new _zipkinTransportHttp.HttpLogger({
+    // endpoint of local docker zipkin instance
+    endpoint: `https://${
+      THEME_OPTIONS.staging ? `staging-` : ``
+    }trace-api.newrelic.com/trace/v1`,
+    headers: {
+      "Api-Key": THEME_OPTIONS.NR_INGEST_KEY,
+      "Data-Format": "zipkin",
+      "Data-Format-Version": 2,
+      options: THEME_OPTIONS,
+      tags: {
+        ...THEME_OPTIONS.traces.tags,
+        ...ciAttributes,
+      },
+    },
+  });
+  recorder = new _zipkin.BatchRecorder({
+    logger,
+    // timeout = 60 hours, must be longer than site's build time
+    timeout: 60 * 60 * 60 * 1000000,
+  });
+  // console.log(recorder)
+  const tracer = new _zipkinJavascriptOpentracing.default({
+    localServiceName: THEME_OPTIONS.SITE_NAME,
+    serviceName: THEME_OPTIONS.SITE_NAME,
+    sampler: new _zipkin.sampler.CountingSampler(1),
+    traceId128Bit: true,
+    recorder,
+    kind: `client`,
+  });
+  return tracer;
+};
+exports.create = create;
+const _processQueue = async () => {
```

Comment:
I think this function is doing a little bit too much. We might want to look at breaking some of the functionality out into their own functions (to better adhere to the [single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle).

Here's a rough sketch of how that _could_ look:

```js
// format an individual trace for logging
const formatTrace = ( trace) => { /* ... */ };

// send the data into new relic
const sendLog = async (queue) => { /* ... */ };

const processQueue = async () => {
  try {
    const queue = logger.queue.map(formatTrace);
    await sendLog(queue);
  } catch(e) {
    // handle errors
  }
};
```

Reply:
OG Typescript: will test 🤠 
but ye some of this is all my additional trace formatting, which we could also abstract 🥳